lib/cpp/src/thrift/server/TNonblockingServer.cpp

```diff
@@ -1105,10 +1113,17 @@ void TNonblockingServer::listenSocket(THRIFT_SOCKET s) {
   serverSocket_ = s;
 
   if (!port_) {
-    sockaddr_in addr;
-    socklen_t size = sizeof(addr);
-    if (!getsockname(serverSocket_, reinterpret_cast<sockaddr*>(&addr), &size)) {
-      listenPort_ = ntohs(addr.sin_port);
+    #ifdef _WIN32
+      struct sockaddr_storage addr;
+      struct sockaddr_in *pAddr = (sockaddr_in *) &addr;
+      socklen_t size = sizeof(addr);
+    #else
+      struct sockaddr_in addr;
+      struct sockaddr_in *pAddr = &addr;
+      socklen_t size = sizeof(addr);
+    #endif
```

Comment:
getsockname fails cause the buffer is not big enough. If you compare the size of "sockaddr_in" and "sockaddr_storage", you'll find out that "sockaddr_storage" has many more fields (due to IPv6 and many other stuff).

Some hints here:

http://stackoverflow.com/questions/14624427/wsaefault-error-when-use-getsockname-function